url29ff2e53f3_nicolas3470_Settlers_of_Catan_AI_tgz-pJ8-robotGameJ8/src/soc/server/SOCServer.java

```diff
@@ -0,0 +1,9048 @@
+                }
+            }
+            else
+            {
+                messageToPlayer(c, gname, "It's not your turn.");
+            }
+        }
+        catch (Exception e)
+        {
+            D.ebugPrintStackTrace(e, "Exception caught");
+        }
+
+        ga.releaseMonitor();
+    }
+
+    /**
+     * Pre-checking already done, end the current player's turn in this game.
+     * Alter game state and send messages to players.
+     * Calls {@link SOCGame#endTurn()}.
+     * On the 6-player board, this may begin the {@link SOCGame#SPECIAL_BUILDING Special Building Phase},
+     * or end a player's placements during that phase.
+     *<P>
+     * Assumes:
+     * <UL>
+     * <LI> ga.canEndTurn already called, to validate player
+     * <LI> ga.takeMonitor already called (not the same as {@link SOCGameList#takeMonitorForGame(String)})
+     * <LI> gamelist.takeMonitorForGame is NOT called, we do NOT have that monitor
+     * </UL>
+     *<P>
+     * As a special case, endTurn is used to begin the Special Building Phase during the
+     * start of a player's own turn, if permitted.  (Added in 1.1.09)
+     *
+     * @param ga Game to end turn
+     * @param pl Current player in <tt>ga</tt>, or null. Not needed except in SPECIAL_BUILDING.
+     *           If null, will be determined within this method.
+     */
+    private void endGameTurn(SOCGame ga, SOCPlayer pl)
+    {
+        final String gname = ga.getName();
+
+        if (ga.getGameState() == SOCGame.SPECIAL_BUILDING)
+        {
+            final int cpn = ga.getCurrentPlayerNumber();
+            if (pl == null)
+                pl = ga.getPlayer(cpn);
+            pl.setAskedSpecialBuild(false);
+            messageToGame(gname, new SOCPlayerElement(gname, cpn, SOCPlayerElement.SET, SOCPlayerElement.ASK_SPECIAL_BUILD, 0));
+        }
+
+        boolean hadBoardResetRequest = (-1 != ga.getResetVoteRequester());
+
+        /**
+         * End the Turn:
+         */
+
+        ga.endTurn();  // May set state to OVER, if new player has enough points to win.
+                       // May begin or continue the Special Building Phase.
+        
+        // If a winner, write stats for each player
+        if(ga.getGameState() == SOCGame.OVER) {
+        	try {
+            	// TODO: write more stats
+        		System.out.println();
+        		System.out.println();
+        		System.out.println("GAME OVER");
+        		System.out.println();
+        		System.out.println();
+        		int[] scores = new int[ga.maxPlayers];
+                boolean[] isRobot = new boolean[ga.maxPlayers];
+                for (int i = 0; i < ga.maxPlayers; ++i)
+                {
+                    scores[i] = ga.getPlayer(i).getTotalVP();
+                    isRobot[i] = ga.getPlayer(i).isRobot();
+                }
+                
+            	String stats = "";
+				FileWriter out_stream = new FileWriter("output.txt",true);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *out\_stream*. The resource is referenced at lines 5472, 5482, 5485, 5488, 5491, 5496, 5499. The resource is closed at line 5501. There are other execution paths that do not contain closure statements, for example, when Writer.write throws an exception. Ensure *out\_stream* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).